### PR TITLE
fix: exclusive CPUs removed from deleted Pod and put back in the defaultCPUSet

### LINKED ISSUE #107074: Exclusive CPUs not removed from deleted Pod and put back in the defaultCPUSet.
### What happened?

The `CPUManager` has logic to periodically cleanup stale state and reclaim exclusive CPUs from pods that have recently terminated. It does this by querying the system for a list of `activePods()` and reclaiming CPUs from any pods it is tracking that are not in this list.

This works fine for most pods, but special care needs to be taken to ensure that CPUs are not accidentally reclaimed from pods that have not yet started yet. Allocation of CPUs to the containers of a pod happens during pod admission (i.e. before the pod is added to the `activePods()` list), so a simple state variable (`pendingAdmissionPod`) is used to indicate which pod is currently being admitted and exclude it from cleanup. Since pod admission is serialized, only one pod will ever be pending admission at a given time, and only a single variable is necessary to track this (i.e. whenever a new pod enters the admission loop to have exclusive CPUs granted to it, `pendingAdmissionPod` is overwritten to point to the new pod, clearing way for the previous one to have its state cleaned up when appropriate).

Unfortunately, this simple procedure can cause problems because `pendingAdmissionPod` is never reset to nil after the `last` pod is admitted. This is usually fine, because the next time a pod comes in for admission, it will be overwritten to point to the new pod. But if no new pods come in, then it continues to point to the last pod that we attempted to admit (forever essentially), making it so that we can never cleanup its state if it gets deleted at some point in the future (because it is always treated as an "active" pod so long as it is pointed to by `pendingAdmissionPod`).

I don't think this issue is critical, since in any practical setting pods will be started and stopped all the time, clearing way for the state of previously admitted pods to be cleaned up. But we should consider a better method of tracking pods that are `notYetActiveButPendingAdmission()` so that we can eliminate this weird edge case.

### What did you expect to happen?

With `--cpu-manager-policy=static` enabled on a node.

Look at the CPU set assigned to one of the system pods running on that node in a non-Guaranteed QOS class:
```
$ kubectl exec -it <pod> -- taskset -cp 1
pid 1's current affinity list: 0-255
```

Create a pod requesting exclusive CPUs:
```
$ cat pod.yaml
apiVersion: v1
kind: Pod
metadata:
  name: test-pod
spec:
  restartPolicy: Never
  containers:
  - image: ubuntu:20.04
    name: test-ctr
    command: ["bash", "-c"]
    args: ["sleep 99999"]
    resources:
      limits:
        cpu: 4000m
        memory: 1Gi
```
```
$ kubectl apply -f pod.yml
pod/test-pod created
```

Look at the set of exclusive CPUs granted to it (we should see 4):
```
$ kubectl exec -it test-pod -- taskset -cp 1
pid 1's current affinity list: 1,2,129,130
```

Look again at the set of CPU set assigned to the pod in the non-Guaranteed QOS class (the 4 CPUs from above are gone):
```
$ kubectl exec -it <pod> -- taskset -cp 1
pid 1's current affinity list: 0,3-128,131-255
```

Delete the test pod:
```
$ kubectl delete pod test-pod
pod "test-pod" deleted
```

Look again at the set of CPU set assigned to the pod in the non-Guaranteed QOS class (we are back to the original set):
```
$ kubectl exec -it <pod> -- taskset -cp 1
pid 1's current affinity list: 0-255
```

### How can we reproduce it (as minimally and precisely as possible)?

Go through the steps above, but the last step doesn't show the original CPU set restored.
```
$ kubectl exec -it <pod> -- taskset -cp 1
pid 1's current affinity list: 0,3-128,131-255
```

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.2", GitCommit:"8b5a19147530eaac9476b0ab82980b4088bbc1b2", GitTreeState:"clean", BuildDate:"2021-09-15T21:38:50Z", GoVersion:"go1.16.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.2", GitCommit:"8b5a19147530eaac9476b0ab82980b4088bbc1b2", GitTreeState:"clean", BuildDate:"2021-09-15T21:32:41Z", GoVersion:"go1.16.8", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
Baremetal
</details>

### OS version

_No response_

### Install tools

_No response_

### Container runtime (CRI) and and version (if applicable)

_No response_

### Related plugins (CNI, CSI, ...) and versions (if applicable)

_No response_

### PR DESCRIPTION
**Pull request by likakuli**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Exclusive CPUs not removed from deleted Pod and put back in the defaultCPUSet.  This will lead to some burstable or besteffort pod can't use spare cpus.
#### Which issue(s) this PR fixes:

Fixes #107074

#### Special notes for your reviewer:
This bug is caused by the pr https://github.com/kubernetes/kubernetes/pull/103979 which is used to fix the issue https://github.com/kubernetes/kubernetes/issues/103952. The root cause is that the result of `GetActivePods` not contains pods in admission stage. So it add a `pendingAdmissionPod` to record it and use it in `removeStaleState`.  

There is another pr #104577 which fixed the problem too and the result of `GetActivePods` will contains pods in admission stage.
#### Does this PR introduce a user-facing change?

```release-note
Release CPU from cpu_manager_state immediately when pod removed successfully.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by fanhaouu** [comment]
LGTM

**Comment by fanhaouu** [comment]
NB

**Comment by likakuli** [comment]
kindly ping @smarterclayton, can you please help review this thx

**Comment by ffromani** [comment]
hi @likakuli thanks for this PR! Could you maybe please elaborate a bit why you want to revert https://github.com/kubernetes/kubernetes/pull/103979 ? Do we know the original issue #103979 fixed is no longer relevant?

**Comment by likakuli** [comment]
> Do we know the original issue #103979 fixed is no longer relevant?

https://github.com/kubernetes/kubernetes/pull/103979#issuecomment-904836223, the root cause of #103979 bas been fixed by https://github.com/kubernetes/kubernetes/pull/104577 already.

**Comment by ffromani** [comment]
Is not obvious to me why, though. The scenario outlined in https://github.com/kubernetes/kubernetes/issues/103952 seems still relevant to me, because while a pod is being admitted, its container cannot be active yet, it's just too early still. I'll do my homework and review again the related changes, though

**Comment by likakuli** [comment]
> I'll do my homework and review again the related changes, though

thx. After https://github.com/kubernetes/kubernetes/pull/104577, the result of `activePods` will contains pods in admission stage.

**Comment by ffromani** [comment]
right, this should be it:
```
Second, the previous check introduced with the pod worker lifecycle
ownership changes was subtly wrong for the admission use case.
Admission has to include pods that haven't yet hit the pod worker,
which CouldHaveRunningContainers was filtering out (because the
pod worker hasn't seen them). Introduce a weaker check -
IsPodKnownTerminated() - that returns true only if the pod is in
a known terminated state (no running containers AND known to pod
worker). This weaker check may only be called from components that
need admitted pods, not other kubelet subsystems.
```

**Comment by ffromani** [comment]
/test pull-kubernetes-node-kubelet-serial-cpu-manager

thanks for the patient explanation. I think you are likely right. Let's however make sure the original issue stay fixed, at very least ensuring e2e tests don't regress (extending them if needed)

**Comment by likakuli** [comment]
kindly ping @ffromani

**Comment by ffromani** [comment]
I'm more and more convinced the change is good.
The serial test outcome looks good. Need to run the other lanes and do verify we have sufficient coverage to prevent regressions resurfacing.

**Comment by likakuli** [comment]
Is it possible to involve the old pr authors to verify this? 

I think this is ok and i have also tested it by myself.

**Comment by ffromani** [comment]
> Is it possible to involve the old pr authors to verify this?

We can tag @cynepco3hahue but I'm afraid they moved to different projects.

> I think this is ok and i have also tested it by myself.

Sure, but the projects needs to be self-sustaining as much as possible, so let's make sure the tests are sufficient - and passing. I'll keep working at quite high priority in this issue and I think we're in a good track to merge.
OFC nothing prevents other contributors to LGTM earlier if they feel so.

**Comment by ffromani** [comment]
neither failures are critical for this PR (actually not related).
xref: https://github.com/kubernetes/kubernetes/issues/120725

**Comment by swatisehgal** [comment]
/test pull-kubernetes-node-kubelet-serial-topology-manager
https://github.com/kubernetes/kubernetes/pull/120773 is now merged so let's see if that helps.

**Comment by likakuli** [comment]
seems timeout also. do i need to rebase this pr with master branch before retest?

**Comment by ffromani** [comment]
> seems timeout also. do i need to rebase this pr with master branch before retest?

we're working on fixing the lanes (esp. the topology manager ones) and having green runs is nice, but nothing suggests the issues are caused by this PR. Atm no action is needed here.

**Comment by likakuli** [comment]
kindly ping @klueska, could you please help to review this, thx

**Comment by likakuli** [comment]
kindly ping @smarterclayton, could you please help to review this pr, thx

**Comment by likakuli** [comment]
any update?

**Comment by ffromani** [comment]
the only update I can provide is that "serial-hugepages" failure is not relevant, and that the topology manager failure should be fixed thanks to https://github.com/kubernetes/kubernetes/pull/121442 . If you resubmit, I'll re-add the LGTM.

**Comment by pacoxu** [comment]
A release note may be needed. 
And this needs a backport. @SergeyKanzhelev @bart0sh

**Comment by ffromani** [comment]
/lgtm

it seems last change was a rebase. Same caveats per https://github.com/kubernetes/kubernetes/pull/120661#issuecomment-1723386569 apply

**Comment by ffromani** [comment]
/remove-lifecycle stale

still relevant

**Comment by likakuli** [comment]
what can i do to push forward this pr?

**Comment by ffromani** [comment]
> what can i do to push forward this pr?

we need approver review, let's try to gently ping on community channels

**Comment by ffromani** [comment]
/lgtm

seems like a rebase

**Comment by pacoxu** [comment]
/assign @mrunalp @klueska @SergeyKanzhelev  
for approval

**Comment by ffromani** [comment]
ping @SergeyKanzhelev @mrunalp @klueska

**Comment by ffromani** [comment]
ping

**Comment by SergeyKanzhelev** [comment]
A few comments:

1. Release note: can you please update in a form of what was the bug and what will be the new behavior. So it will be clear for people reading it what was fixed.
2. Is priority accurate? Is there any need to cherry-pick this fix? Or the bug was there for too long to justify the priority bump?
3. Can any test be added that demonstrates the bug be added? Perhaps some e2e test that will be able to catch a possible regression here. Right now I only see unit tests, which I don't think demonstrate the bug and rather validate some internal details.

**Comment by ffromani** [comment]
>     2. Is priority accurate? Is there any need to cherry-pick this fix? Or the bug was there for too long to justify the priority bump?

I'm unsure between this priority and `backlog`. I don't think this warrants a backport; I have investigated some user-affected flows (also in RH) and the impacted flows are fairly well defined and relatively rare.
And yes, this bug is present since quite some timew.

>      3. Can any test be added that demonstrates the bug be added? Perhaps some e2e test that will be able to catch a possible regression here. Right now I only see unit tests, which I don't think demonstrate the bug and rather validate some internal details.

This is a good point, I like the direction about adding more e2e tests. I think there's room to crete a valid e2e test and I can help with the details.

**Comment by ffromani** [comment]
> > ```
> > 2. Is priority accurate? Is there any need to cherry-pick this fix? Or the bug was there for too long to justify the priority bump?
> > ```
> 
> I'm unsure between this priority and `backlog`. I don't think this warrants a backport; I have investigated some user-affected flows (also in RH) and the impacted flows are fairly well defined and relatively rare. And yes, this bug is present since quite some timew.
> 
> > ```
> >  3. Can any test be added that demonstrates the bug be added? Perhaps some e2e test that will be able to catch a possible regression here. Right now I only see unit tests, which I don't think demonstrate the bug and rather validate some internal details.
> > ```
> 
> This is a good point, I like the direction about adding more e2e tests. I think there's room to crete a valid e2e test and I can help with the details.

I got a test which proves the point by side effect in https://github.com/kubernetes/kubernetes/pull/127506. I'm not sure is worth a more targeted test, but I'm open to suggestions.

**Comment by ffromani** [comment]
@likakuli howdy! we just miss a e2e test here. Can you please maybe borrow this code from me https://github.com/kubernetes/kubernetes/pull/127506/files#diff-897688174d4d6c5ec8b4080a2bbeeddc9e5442781586de6154a32b0b80cfc336R199 and add to your PR?
Or, if you prefer, feel free to reimplement any test to mine. That should be sufficient to demonstrate the bug and the fix.

**Comment by likakuli** [comment]
> @likakuli howdy! we just miss a e2e test here. Can you please maybe borrow this code from me https://github.com/kubernetes/kubernetes/pull/127506/files#diff-897688174d4d6c5ec8b4080a2bbeeddc9e5442781586de6154a32b0b80cfc336R199 and add to your PR? Or, if you prefer, feel free to reimplement any test to mine. That should be sufficient to demonstrate the bug and the fix.

Thanks! I will do it after holiday

**Comment by ffromani** [comment]
> > @likakuli howdy! we just miss a e2e test here. Can you please maybe borrow this code from me https://github.com/kubernetes/kubernetes/pull/127506/files#diff-897688174d4d6c5ec8b4080a2bbeeddc9e5442781586de6154a32b0b80cfc336R199 and add to your PR? Or, if you prefer, feel free to reimplement any test to mine. That should be sufficient to demonstrate the bug and the fix.
> 
> Thanks! I will do it after holiday

cool, thanks! Meanwhile I'll work a bit on extracting my own test and validating, so you can easily use it or not as you see fit.

**Comment by ffromani** [comment]
ok, good news: #120661 fixes the test I added in https://github.com/kubernetes/kubernetes/pull/127506/files#diff-897688174d4d6c5ec8b4080a2bbeeddc9e5442781586de6154a32b0b80cfc336R199 , hence the latter (or an amended version of) can be used to validate the former

**Comment by SergeyKanzhelev** [comment]
@ffromani @likakuli let me know how you want to proceed on those PRs in terms of ordering

**Review by ffromani — CHANGES_REQUESTED** [review]
mind adding `Co-authored-by: Francesco Romani <fromani@redhat.com>`  before/after the `Signed-off-by` line in the test commit message?

**Comment by ffromani** [comment]
the test failure is real, investigating

**Comment by ffromani** [comment]
> the test failure is real, investigating

ok, silly me: the test can't work till we actually add the metric, which is the whole point of my PR in the first place. At this point is probably easier to cherry-pick the fix from @likakuli in my own PR :\

**Comment by likakuli** [comment]
> ok, silly me: the test can't work till we actually add the metric, which is the whole point of my PR in the first place. At this point is probably easier to cherry-pick the fix from @likakuli in my own PR :\

haha,can cherry-pick this PR If needed

**Comment by ffromani** [comment]
> > ok, silly me: the test can't work till we actually add the metric, which is the whole point of my PR in the first place. At this point is probably easier to cherry-pick the fix from @likakuli in my own PR :\
> 
> haha,can cherry-pick this PR If needed

sure, done! https://github.com/kubernetes/kubernetes/pull/127506#issuecomment-2399543876

**Comment by haircommander** [comment]
this looks to be done in https://github.com/kubernetes/kubernetes/pull/127506 should we close this?

**Comment by ffromani** [comment]
/close

reason: https://github.com/kubernetes/kubernetes/pull/120661#issuecomment-2399530498

**Pull request closed without merging**
